File: src/lib/market-calendar.ts
/**
 * Market calendar utility — frontend mirror of backend/market_calendar.py.
 *
 * Keeps UI awareness of trading-day vs holiday vs early-close in sync
 * with the backend so the Health page does not raise false alarms
 * outside US stock-market hours.
 *
 * Update annually when CBOE publishes the next year's schedule.
 */

// 2026 market holidays — full close (YYYY-MM-DD strings for cheap comparison).
const MARKET_HOLIDAYS_2026 = new Set<string>([
    '2026-01-01', // New Year's Day
    '2026-01-19', // MLK Day
    '2026-02-16', // Presidents' Day
    '2026-04-03', // Good Friday
    '2026-05-25', // Memorial Day
    '2026-06-19', // Juneteenth
    '2026-07-03', // Independence Day (observed)
    '2026-09-07', // Labor Day
    '2026-11-26', // Thanksgiving
    '2026-12-25', // Christmas
]);

// 2026 early close days — market closes at 1:00 PM ET.
const EARLY_CLOSE_2026 = new Set<string>([
    '2026-07-02', // Day before Independence Day
    '2026-11-27', // Black Friday
    '2026-12-24', // Christmas Eve
]);

const MARKET_OPEN_MINUTES = 9 * 60 + 30; // 09:30 ET
const NORMAL_CLOSE_MINUTES = 16 * 60;    // 16:00 ET
const EARLY_CLOSE_MINUTES = 13 * 60;     // 13:00 ET

/**
 * Convert any Date into a synthetic Date whose `getFullYear/Month/Date/
 * Day/Hours/Minutes` fields read out the wall-clock time in America/
 * New_York. The resulting Date's underlying UTC instant is meaningless
 * — only call the local accessors on it.
 */
function toETWallClock(now: Date = new Date()): Date {
    // toLocaleString('en-US', { timeZone: 'America/New_York' }) renders
    // the ET wall clock as a string; re-parsing that string with
    // `new Date(...)` treats it as the user's local time, which means
    // the local accessors then return the ET components we want.
    return new Date(
        now.toLocaleString('en-US', { timeZone: 'America/New_York' }),
    );
}

function toDateString(d: Date): string {
    const yyyy = d.getFullYear();
    const mm = String(d.getMonth() + 1).padStart(2, '0');
    const dd = String(d.getDate()).padStart(2, '0');
    return `${yyyy}-${mm}-${dd}`;
}

export function isMarketDay(now: Date = new Date()): boolean {
    const et = toETWallClock(now);
    const dow = et.getDay(); // 0=Sun, 6=Sat
    if (dow === 0 || dow === 6) return false;
    return !MARKET_HOLIDAYS_2026.has(toDateString(et));
}

/** Hour at which the market closes today (16 normal, 13 early close). */
export function getMarketCloseHour(now: Date = new Date()): number {
    const et = toETWallClock(now);
    return EARLY_CLOSE_2026.has(toDateString(et)) ? 13 : 16;
}

/**
 * True only if RIGHT NOW the US stock market is open.
 *
 * Honors:
 *   - weekends (Sat/Sun) → false
 *   - holidays           → false
 *   - pre-market (< 09:30 ET) → false
 *   - post-close (≥ 16:00 ET, or ≥ 13:00 ET on early-close days) → false
 */
export function isMarketOpen(now: Date = new Date()): boolean {
    if (!isMarketDay(now)) return false;
    const et = toETWallClock(now);
    const minutes = et.getHours() * 60 + et.getMinutes();
    const closeMinutes = EARLY_CLOSE_2026.has(toDateString(et))
        ? EARLY_CLOSE_MINUTES
        : NORMAL_CLOSE_MINUTES;
    return minutes >= MARKET_OPEN_MINUTES && minutes < closeMinutes;
}

/**
 * Returns a one-word label for the current market state.
 * Prefers Weekend/Holiday over Closed when applicable so the UI
 * can explain why the market isn't open.
 */
export function getMarketStatusLabel(now: Date = new Date()): string {
    if (!isMarketDay(now)) {
        const et = toETWallClock(now);
        const dow = et.getDay();
        if (dow === 0 || dow === 6) return 'Weekend';
        return 'Holiday';
    }
    return isMarketOpen(now) ? 'Open' : 'Closed';
}

File: backend/market_calendar.py
"""
Market calendar utility — holiday and early-close awareness.

Used by health status writers, time stops, and position monitor to
determine whether the market is currently open and when it closes.

US stock market (NYSE/CBOE) 2026 holidays and early closes. Update
this file annually or when CBOE publishes the next year's schedule.

Standalone — depends on nothing else in the trading system.
"""

from __future__ import annotations

from datetime import date, datetime, time, timedelta
from zoneinfo import ZoneInfo

ET = ZoneInfo("America/New_York")

# 2026 Market Holidays (full close)
MARKET_HOLIDAYS_2026: frozenset[date] = frozenset([
    date(2026, 1, 1),    # New Year's Day
    date(2026, 1, 19),   # Martin Luther King Jr. Day
    date(2026, 2, 16),   # Presidents' Day
    date(2026, 4, 3),    # Good Friday
    date(2026, 5, 25),   # Memorial Day
    date(2026, 6, 19),   # Juneteenth
    date(2026, 7, 3),    # Independence Day (observed)
    date(2026, 9, 7),    # Labor Day
    date(2026, 11, 26),  # Thanksgiving Day
    date(2026, 12, 25),  # Christmas Day
])

# 2026 Early Close Days (market closes at 1:00 PM ET)
EARLY_CLOSE_2026: frozenset[date] = frozenset([
    date(2026, 7, 2),    # Day before Independence Day
    date(2026, 11, 27),  # Day after Thanksgiving (Black Friday)
    date(2026, 12, 24),  # Christmas Eve
])

# Standard market hours
MARKET_OPEN = time(9, 30, 0)
MARKET_CLOSE = time(16, 0, 0)
EARLY_CLOSE = time(13, 0, 0)

# Time stop offsets (close positions before market close)
TIME_STOP_SHORT_GAMMA_OFFSET = timedelta(minutes=90)    # 2:30 PM on normal days
TIME_STOP_ALL_POSITIONS_OFFSET = timedelta(minutes=15)  # 3:45 PM on normal days


def is_market_day(check_date: date | None = None) -> bool:
    """True if the given date is a trading day (not weekend or holiday)."""
    d = check_date or date.today()
    if d.weekday() >= 5:  # Saturday=5, Sunday=6
        return False
    if d in MARKET_HOLIDAYS_2026:
        return False
    return True


def get_market_close_time(check_date: date | None = None) -> time:
    """Return the market close time for the given date (early close aware)."""
    d = check_date or date.today()
    if d in EARLY_CLOSE_2026:
        return EARLY_CLOSE
    return MARKET_CLOSE


def is_market_open(now: datetime | None = None) -> bool:
    """
    True if the US stock market is currently open.

    `now` may be naive (interpreted as ET) or timezone-aware.
    Honors weekends, holidays, and early closes.
    """
    if now is None:
        now = datetime.now(ET)
    elif now.tzinfo is None:
        now = now.replace(tzinfo=ET)

    now_et = now.astimezone(ET)
    today = now_et.date()

    if not is_market_day(today):
        return False

    close = get_market_close_time(today)
    current_time = now_et.time()
    return MARKET_OPEN <= current_time < close


def get_time_stop_230pm(check_date: date | None = None) -> time:
    """
    Short-gamma time stop: 90 minutes before market close.
    Normal days: 2:30 PM ET. Early close: 11:30 AM ET.
    """
    d = check_date or date.today()
    close = get_market_close_time(d)
    close_dt = datetime.combine(d, close, tzinfo=ET)
    stop_dt = close_dt - TIME_STOP_SHORT_GAMMA_OFFSET
    return stop_dt.time()


def get_time_stop_345pm(check_date: date | None = None) -> time:
    """
    All-positions time stop: 15 minutes before market close.
    Normal days: 3:45 PM ET. Early close: 12:45 PM ET.
    """
    d = check_date or date.today()
    close = get_market_close_time(d)
    close_dt = datetime.combine(d, close, tzinfo=ET)
    stop_dt = close_dt - TIME_STOP_ALL_POSITIONS_OFFSET
    return stop_dt.time()
